lib/propolis/src/block/attachment.rs

```diff
@@ -66,15 +66,20 @@ impl BlockData {
                 }
                 Err(err) => match err {
                     ReqError::NonePending | ReqError::Paused => {
-                        let guard = self.lock.lock().unwrap();
                         // Double-check for attachment-related "error"
                         // conditions under protection of the lock before
                         // finally blocking.
-                        if check_state(att_state).is_err() {
-                            return None;
+                        let guard = self.lock.lock().unwrap();
+                        match check_state(att_state) {
+                            Err(ReqError::Stopped | ReqError::Detached) => {
+                                return None;
+                            }
+                            Ok(())
+                            | Err(ReqError::Paused | ReqError::NonePending) => {
+                                let _guard = self.cv.wait(guard).unwrap();
+                                continue;
+                            }
                         }
```

Comment:
I would just make the calls to `is_attached()` and `is_stopped()` here, instead of using `check_state`.  Handling the `NonePending` error case is misleading, since `check_state` never emits it.  It suggest that more is being checked than really is.

Reply:
Good call, I'll fix this up.

Reply:
Should be fixed in 099e2be.